cmd/dagger/module.go

```diff
@@ -507,88 +512,157 @@ This command is idempotent: you can run it at any time, any number of times. It
 			if err != nil {
 				return fmt.Errorf("failed to get source root subpath: %w", err)
 			}
-			srcRootAbsPath := filepath.Join(contextDirPath, srcRootSubPath)
+			baseSrcRootPath := filepath.Join(contextDirPath, srcRootSubPath)
 
-			if engineVersion := getCompatVersion(); engineVersion != "" {
-				modSrc = modSrc.WithEngineVersion(engineVersion)
-			}
-
-			modSDK, err := modSrc.SDK().Source(ctx)
-			if err != nil {
-				return fmt.Errorf("failed to get module SDK: %w", err)
-			}
-			if developSDK != "" {
-				if modSDK != "" && modSDK != developSDK {
-					return fmt.Errorf("cannot update module SDK that has already been set to %q", modSDK)
+			modSrcs := make(map[string]*dagger.ModuleSource)
+			if developRecursive {
+				ctx, span := Tracer().Start(ctx, "load module", telemetry.Encapsulate())
+				err := collectLocalModulesRecursive(ctx, modSrc, modSrcs)
+				telemetry.End(span, func() error { return err })
+				if err != nil {
+					return err
 				}
-				modSDK = developSDK
-				modSrc = modSrc.WithSDK(modSDK)
+			} else {
+				modSrcs[baseSrcRootPath] = modSrc
 			}
 
-			modSourcePath, err := modSrc.SourceSubpath(ctx)
-			if err != nil {
-				return fmt.Errorf("failed to get module source subpath: %w", err)
-			}
-			// if SDK is set but source path isn't and the user didn't provide --source, we'll use the default source path
-			if modSDK != "" && modSourcePath == "" && developSourcePath == "" {
-				inferredSourcePath, err := inferSourcePathDir(srcRootAbsPath)
-				if err != nil {
-					return err
+			ctx, span := Tracer().Start(ctx, "develop")
+			defer telemetry.End(span, func() error { return err })
+
+			eg, ctx := errgroup.WithContext(ctx)
```

Comment:
hey look at that, it's the (elusive) correct use case for errgroup.WithContext: cancel the whole thing when the first one fails.